flasc/energy_ratio/energy_ratio.py

```diff
@@ -105,32 +105,47 @@ def _compute_energy_ratio_single(df_,
     df_ = add_ws_bin(df_, ws_cols, ws_step, ws_min, ws_max, remove_all_nulls=remove_all_nulls)
     df_ = add_wd_bin(df_, wd_cols, wd_step, wd_min, wd_max, remove_all_nulls=remove_all_nulls)
 
-    
-
     # Assign the reference and test power columns
     df_ = add_power_ref(df_, ref_cols)
     df_ = add_power_test(df_, test_cols)
 
     bin_cols_without_df_name = [c for c in bin_cols_in if c != 'df_name']
     bin_cols_with_df_name = bin_cols_without_df_name + ['df_name']
 
-    
-    
+    # Group df_
     df_ = (df_
         .filter(pl.all_horizontal(pl.col(bin_cols_with_df_name).is_not_null())) # Select for all bin cols present
         .groupby(bin_cols_with_df_name, maintain_order=True)
-        .agg([pl.mean("pow_ref"), pl.mean("pow_test"),pl.count()]) 
-        .with_columns(
-            [
-                # Get the weighting by counts
-                pl.col('count').min().over(bin_cols_without_df_name).alias('count_weight') if weight_by == 'min' else
-                pl.col('count').sum().over(bin_cols_without_df_name).alias('count_weight')
-            ]
+        .agg([pl.mean("pow_ref"), pl.mean("pow_test"),pl.count()])
+        )
+    
+    # Determine the weighting of the ws/wd bins
+
+    if df_freq_pl is None:
+        # Determine the weights per bin as either the min or sum count
+        df_ = (df_
+            .with_columns(
+                [
+                    # Get the weighting by counts
+                    pl.col('count').min().over(bin_cols_without_df_name).alias('weight') if weight_by == 'min' else
+                    pl.col('count').sum().over(bin_cols_without_df_name).alias('weight')
+                ]
+            )
         )
+    
+    else:
+        # Use the weights in df_freq_pl directly
+        df_ = (df_.join(df_freq_pl, on=['wd_bin','ws_bin'], how='left')
+              .with_columns(pl.col('weight').fill_null(strategy="zero"))
+        )
+
```

Comment:
@ejsimley, reviewing your comments made me realize a possible issue.  Say the weights coming from df_freq assign weight to a bin for which one of the cases has 0 points.  In this case at least we need to remove the bin right?  I can make a test for this case, maybe this happens anyway and I'm not realizing it, but maybe we also need to explicitly check it right?

Reply:
I think you're right, but now I'm wondering if assigning a weight from df_freq to a bin where only one of the cases has data is any different than if we used the "sum" weighting option for the same scenario. Either way, there would be a non-zero weight applied to a bin with data in only one case. I guess that wound't make sense when computing the uplift for two cases and it would be best to exclude the bin entirely. It should be ok if a user is only interested in the individual energy ratios, but we could assume that if someone's not interested in computing the uplift, they'd just create energy ratios with only one case at a time.

So I vote for setting the weight to zero if data are missing for any of the cases in a particular bin, though not sure the best way to do that.

Reply:
This is a good catch, I just pushed up a fix that ensures that all ws/wd bins appear in all dataframes, and added a test that this works as expected